Python/bytecodes.c

```diff
@@ -3829,14 +3829,32 @@ dummy_func(
             PyFunctionObject *func_obj = (PyFunctionObject *)
                 PyFunction_New(codeobj, GLOBALS());
 
-            Py_DECREF(codeobj);
             if (func_obj == NULL) {
                 GOTO_ERROR(error);
             }
-
-            _PyFunction_SetVersion(
-                func_obj, ((PyCodeObject *)codeobj)->co_version);
+            PyCodeObject *code = (PyCodeObject *)codeobj;
+            if (func_obj->func_builtins != tstate->interp->builtins) {
+                _PyFunction_SetVersion(func_obj, 0);
+            }
+            else if (code->co_version == 0) {
+                code->co_version = tstate->interp->func_state.next_version;
+                if (tstate->interp->func_state.next_version != 0) {
+                    tstate->interp->func_state.next_version++;
+                }
+                assert(code->co_expected_globals_version == 0);
+                if (PyDict_CheckExact(GLOBALS())) {
+                    code->co_expected_globals_version = ((PyDictObject *)GLOBALS())->ma_version_tag;
+                }
+                _PyFunction_SetVersion(func_obj, code->co_version);
+            }
+            else {
+                if (PyDict_CheckExact(GLOBALS()) &&
+                    code->co_expected_globals_version == ((PyDictObject *)GLOBALS())->ma_version_tag) {
+                    _PyFunction_SetVersion(func_obj, code->co_version);
+                }
+            }
             func = (PyObject *)func_obj;
+            Py_DECREF(codeobj);
```

Comment:
This feels like an awfully large block of code for the JIT. Maybe it should be moved to a helper function?

Also, my brain is too small to fully grasp what will happen if we have a local function that references a global and is repeatedly created (maybe a generator expression inside a hot loop), when the referenced global changes (or some other global, given how dict versions work IIRC). Is that going to completely disable tier 2 for that hot loop, or is it just going to use `_LOAD_GLOBAL_MODULE` instead of `_LOAD_CONST_INLINE`?

A variant of the new test may reveal what happens more easily than trying to reason it through...

Reply:
> This feels like an awfully large block of code for the JIT. Maybe it should be moved to a helper function?

Maybe, but that is the responsibility of the JIT builder. Let's not complicate the interpreter specification based on the current behavior of the JIT.

> Is that going to completely disable tier 2 for that hot loop?

If the globals are changed, then we will get a different function version number.
That will disable some tier 2 optimizations. There is no reason why it should completely disable tier 2 optimization.

Reply:
I would like to play with this to see what actually happens. As I said, my brain is too small to reason it through, and your response ("should") does not allay my worries.